Examples

juliendehos opened the pull request:
Hi,
Thanks for developing this project.
This PR adds some examples that may help for #8. The spaceinvaders example is very similar to https://gitlab.com/juliendehos/haskell-invaders .

Closes #8

Commit:
shine-examples (spaceinvaders)

Commit:
examples: add mountaincar + update spaceinvaders

Commit:
examples: update readme

fgaz commented:
Thanks for this!

I'm pretty busy atm, I'll do a proper review in a week or so

fgaz commented on shine-examples/README.md:
````diff
@@ -0,0 +1,15 @@
+
+```
+chromium dist/build/shine-spaceinvaders/shine-spaceinvaders.jsexe/index.html
+chromium dist/build/shine-mountaincar/shine-mountaincar.jsexe/index.html
````
Please use a generic "`$BROWSER`"

fgaz commented on shine-examples/nixpkgs.nix:
```diff
@@ -0,0 +1,56 @@
+  nixpkgs-src = bootstrap.fetchFromGitHub {
+    owner = "NixOS";
+    repo  = "nixpkgs";
+    rev = "725b5499b89fe80d7cfbb00bd3c140a73cbdd97f";
```
I don't think we should tie the user to nix, and especially to a particular nixpkgs.
This also requires updating from time to time (I ran the command and it started compiling ghcjs and ran out of memory, but maybe that's normal, I'm used to reflex's cache).

juliendehos replied:
Ok. I reconfigured the `shine-examples` project for using `reflex-platform` (as explained in its Project Development doc), and I updated the readme. Using the reflex's cache, the project should now be much faster to build.

fgaz commented on shine-examples/mountaincar/Main.hs:
```diff
@@ -0,0 +1,67 @@
+eventH _ g = g
+
+idleH :: Double -> G.Game -> G.Game
+idleH _ = G.step
```
Are you ignoring the time delta to simplify the game, to avoid glitches, or for other reasons? Could you write it in a comment?

juliendehos replied:
The mountain car problem has a fixed time step, generally. I added a comment in the code to explain that.

fgaz commented on shine-examples/shine-examples.cabal:
```diff
@@ -0,0 +1,32 @@
+    hs-source-dirs:     mountaincar
+    default-language:   Haskell2010
+    ghc-options:        -O2 -threaded -rtsopts -with-rtsopts=-N -Wall 
+                        -Wno-unused-top-binds -Wno-missing-signatures -Wno-missing-home-modules
```
* Is `-O2` necessary?
* Does the `-N` actually do anything in ghcjs?
* `-Wall` is good :)
* I'm very much against disabling those warnings, except maybe missing-signatures.

edit: wait, why does this trigger -Wmissing-home-modules? Is it a ghcjs/cabal bug?

juliendehos replied:
Ok, I've updated the options.
Indeed, ghcjs seems to ignore the `other-modules` field, hence the `-Wno-missing-home-modules` option.

fgaz reviewed (commented):
> I'll do a proper review in a week or so

Well, I should just stop giving estimates altogether :D 

---

Looks great!

I left some minor comments

Also please rebase to latest master, I just fixed CI

Commit:
fix fgaz comments:

- switch to reflex-platform
- improve readme
- comment time handling, in mountaincar
- clean cabal files

juliendehos commented:
Thanks for your comments. I've tried to fix them. Let me know if you see some things to improve.

fgaz committed:
Revert "fix fgaz comments:"

This reverts commit c8eb4defc7d8447d765f264cf6ddf9bfa844fdb5.

fgaz committed:
fix fgaz comments:

- make build instructions more unopinionated
- improve readme
- comment time handling, in mountaincar
- clean cabal files

(cherry picked from commit c8eb4defc7d8447d765f264cf6ddf9bfa844fdb5 and
edited)

fgaz commented:
Thank you for the fixes! I reworked it a bit more and I'm going to squash&merge + publish it in a couple of days.

fgaz commented:
And here it is:
https://fgaz.github.io/shine/shine-examples/
https://hackage.haskell.org/package/shine-examples
Thanks again for the contribution!
